Thumbnail plot rendering optimisations

olgamelnichuk opened the pull request:
fixes for http://bar.ebi.ac.uk:8080/trac/ticket/2795 and http://bar.ebi.ac.uk:8080/trac/ticket/3207
- Thumbnail plot generation logic moved from AtlasPlotter to ThumbnailPlot
- The plot scaling code is run on the server-side now
- Added ability to draw thumbnail plot for a given design element accession, not gene id
- Added unit tests 

@alf239 please have a look

Commit:
a copy of thumbnail plot code (from AtlasPlotter) placed into view controller

Commit:
making private fields final..

Commit:
using ternary operator...

Commit:
making private fields final...

Commit:
factor value ordering fixed..

Commit:
added design element accession to the ListResultRowExperiment..
(to be able to draw thumbnail plots by this design element)

Commit:
generating thumbnail plots by design element accession

Commit:
added scale support... do we really want to show NaN values on thumbnail plots?

Commit:
code simplicity

Commit:
typo

Commit:
renaming rangeOf() -> range()

Commit:
TODO left from an old code...

Commit:
Merge remote branch 'origin/develop' into thumbplots

Commit:
code style

Commit:
fixing compilation problem after merge with develop: using ewd.getUniqueEFVs() instead of ewd.getUniqueFactorValues()..

Commit:
code clean up: removed unused thumbnail plotting code from AtlasPlotter

Commit:
java doc added

Commit:
bugfix: latest subset of points was forgotten

Commit:
added ThumbnailPlotTest

Commit:
removed as useless; ThumbnailPlotTest created instead

olgamelnichuk commented:
Need to postpone a bit with merge. I'd like to rename a few classes..

rpetry commented on atlas-web/src/main/java/ae3/model/ListResultRow.java:
```diff
@@ -86,12 +86,10 @@ public class ListResultRow implements Comparable<ListResultRow> {
     public float getMinPval() {
-        if (isMixedCell())
+        if (isMixedCell()) {
             return Math.min(Math.abs(minPval_dn), Math.abs(minPval_up));
```
Do we need Math.abs here? Can pvals be less than 0?

olgamelnichuk replied:
By the theory  p-value could not be negative, but in our code it could be everything;  anyway fixing it is out of scope of this task.

rpetry replied:
I'll make a separate change if this is out of scope, but I'd argue that if we get a negative pvalue anywhere, then we have a problem with our system that needs investigating, rather than hiding behind Math.abs.

Commit:
-added illegal arguments check to simplify semantics of predicates;
-added two more predicates to find ExperimentPart by at-least-one-gene and by array design accession
-eliminated usage of DataPredicate class
-removed DataPredicates class

Commit:
no ambiguity any more :)

Commit:
removed unused method

Commit:
one 'final' forgotten to be removed..

Commit:
added ExperimentPartCriteria unit tests

Commit:
changed inner static class visibility to private

Commit:
using guava's newHashMap() instead of new HashMap(...)

alf239 commented on atlas-data-storage/src/test/java/uk/ac/ebi/gxa/data/ExperimentPartCriteriaTest.java:
```diff
@@ -0,0 +1,251 @@
+        assertEquals(AD2.getAccession(), expPart.getArrayDesign().getAccession());
+
+        expPart = (new ExperimentPartCriteria())
+                .hasArrayDesignAccession(AD1.getAccession())
```
Why not static? At least the starting method, e.g.,

``` java
ExperimentPart expPart = ExperimentPartCriteria.part()
               .hasArrayDesignAccession(AD2.getAccession())
               .apply(ewd);
```

Static import will make it nice-n-shiny

alf239 commented on atlas-data-storage/src/test/java/uk/ac/ebi/gxa/data/ExperimentPartCriteriaTest.java:
```diff
@@ -0,0 +1,251 @@
+        ExperimentWithData ewd = createExperimentWithData();
+
+        ExperimentPart expPart = (new ExperimentPartCriteria())
+                .containsDeAccessions(asList("DE-3"))
```
varargs, perhaps?

alf239 commented on atlas-data-storage/src/test/java/uk/ac/ebi/gxa/data/ExperimentPartCriteriaTest.java:
```diff
@@ -0,0 +1,251 @@
+        ExperimentWithData ewd = createExperimentWithData();
+
+        ExperimentPart expPart = (new ExperimentPartCriteria())
+                .containsEfEfv("EF1", "EFV11")
```
will we ever look for EF-only criteria?

Commit:
method renamed: ExperimentPartCriteria.apply() -> .retrieve()

Commit:
added static create method for ExperimentPartCriteria

Commit:
rename ExperimentPartCriteria.retrieve(...) to .retrieveFrom(...)
